### PR TITLE
Parallelise BoutOutputs and squashoutput

### PR DESCRIPTION
**Pull request by johnomotani**
Uses the `multiprocessing` package to allow creating a set of worker processes which each read a subset of the files to collect from. Should allow a significant speedup on computers with parallel filesystems. ~~There is quite a bit of code more-or-less copied from `collect`. In principle we could refactor to use common utility functions, but I thought it was better not to modify the existing `collect` functionality.~~

~~Not at all well tested yet. In particular handling `FieldPerp`s is quite convoluted and I didn't have output files with `FieldPerp`s handy to check. I did try one set of output files, and got an identical result to the serial version. I also tried squashing some files on Marconi - I didn't wait for the run to finish, but reading `Field3D`s that are \~3GiB went down from \~300s in serial to \~130s using 4 processors.~~

Now includes tests and refactoring of `collect`/`BoutOutputs` code.

### CODE REVIEW

File: boutdata/data.py [hunk 1]
```diff
@@ -955,7 +963,15 @@ def __init__(
...
         caching=False,
         DataFileCaching=True,
-        **kwargs
+        info=False,
+        xguards=True,
+        yguards=False,
+        tind=None,
+        xind=None,
+        yind=None,
+        zind=None,
```
Comment (loeiten) [inline boutdata/data.py:972]: Should these be documented as well, or are these part of input to `_caching_collect()`?
If yes, maybe you could write what are the `kwargs` which are not part of `_caching_collect()`?

File: boutdata/data.py [hunk 2]
```diff
@@ -1034,8 +1131,60 @@ def __init__(
...
+                self._workers.append(
+                    (
+                        Process(
+                            target=self._worker_function,
+                            args=(child_connection, proc_list, self._shared_buffer_raw),
+                        ),
+                        parent_connection,
+                    )
+                )
+                self._workers[-1][0].start()
```
Comment (loeiten) [inline boutdata/data.py:1177]: The constructor is quite long.
Would this be a good time to split it to more (private) sub-routines in order to improve maintainability?
For example it looks to code under `if self._parallel is not False:` could be made into a sub-routine.

File: boutdata/data.py [hunk 3]
```diff
@@ -1034,8 +1131,60 @@ def __init__(
                 self._datacachesize = 0
                 self._datacachemaxsize = self._caching * 1.0e9
 
+        if self._parallel is not False:
```
Comment (loeiten) [inline boutdata/data.py:1134]: Would this be better?

```suggestion
        if self._parallel:
```

File: boutdata/data.py [hunk 4]
```diff
@@ -1241,10 +1390,396 @@ def _collect(self, *args, **kwargs):
...
+            if temp_yindex < 0:
+                # No data for FieldPerp on this processor
+                return None, None
+
+        shared_result[global_slices] = datafile.read(varname, ranges=local_slices)
+
+        if is_fieldperp:
+            return temp_yindex, var_attributes
+
+        return None, None
```
Comment (loeiten) [inline boutdata/data.py:1742]: Maybe this could also be split in several sub-routines?

**Review by loeiten — COMMENTED** [review]
Awesome to have a parallelized routine for this 😄 

I am in favour of refactoring parts from `collect` in order to improve maintainability. In my experience technical debt has a tendency to accelerate once it gets momentum 😐.

File: boutdata/data.py (hunk 3 above)
Comment (johnomotani) [reply boutdata/data.py:1134]: This was deliberate - I wanted to allow `parallel = 0` to be passed as a command-line argument (assuming that would only accept `int`) to mean 'use all available processors'. Not sure that's it's a great interface though... My `bout-squashoutput` command-line executable interface was broken (and I don't have the time to spend debugging it) so I stopped thinking about the command-line arguments.

**Comment by johnomotani** [comment]
@loeiten agreed some refactoring would be nice. If we merge the tests for `collect()` and `squashoutput()` #38, I'm much happier with refactoring `collect()`, etc.

**Comment by johnomotani** [comment]
The parallel branch of squashoutput is now covered by all the squash/collect tests (and several bugs affecting non-default arguments fixed). A bit of refactoring left to do before this is ready to merge.

**Comment by johnomotani** [comment]
I'm trying a bit of refactoring to remove code duplication. I'd like to get rid of this
https://github.com/boutproject/boutdata/blob/a6199d32aedfc804f2180fd89e6bf5a564ff4133/boutdata/collect.py#L685-L690
so I don't have to have `t_array` loaded. That bit of code doesn't seem to me like it should do anything (I think we always write all floats with the same precision from BOUT++) - I'm not sure what the original intention was. @bendudson do you remember?

**Comment by bendudson** [comment]
Hi @johnomotani sorry no I don't know why that's there. It doesn't look like it's doing anything particularly useful. I'm happy for it to be removed

**Comment by ZedThree** [comment]
Here's the original commit: https://github.com/boutproject/BOUT-dev/commit/eb72cd22c61c5ee7bfbda1f77455584dc62ad97a

Looks like it was to workaround an issue when storing using floats instead of doubles.

**Comment by johnomotani** [comment]
Thanks @bendudson @ZedThree. Sounds like it was a work-around for a bug in the netcdf Python libraries at the time. I just checked reading with `netcdf4` and everything seems fine
```
$ ncdump -h test.nc 
netcdf test {
dimensions:
	x = 4 ;
	y = 5 ;
variables:
	double foo(x, y) ;
	float bar(x, y) ;
}
```
```
In [5]: read = Dataset("test.nc", "r")

In [6]: foo = read["foo"][...]

In [7]: foo.dtype
Out[7]: dtype('float64')

In [8]: bar = read["bar"][...]

In [9]: bar.dtype
Out[9]: dtype('float32')
```
so I think we're safe to delete that bit of code.

File: boutdata/data.py [hunk 5]
```diff
@@ -1172,9 +1304,8 @@ def redistribute(self, npes, nxpe=None, mxg=2, myg=2, include_restarts=True):
                         # FieldPerp?
                         # check is not perfect, fails if ny=nz
                         raise ValueError(
-                            "Error: Found FieldPerp '"
-                            + v
-                            + "'. This case is not currently handled by BoutOutputs.redistribute()."
+                            "Error: Found FieldPerp '{}'. This case is not currently "
```
Comment (ZedThree) [inline boutdata/data.py:1307]: Probably not something for this PR, but `restart.redistribute` can handle `FieldPerp`s -- much of the logic is repeated between `BoutOutputs.redistribute` and the version in `restart`, could we merge them somehow?

**Review by ZedThree — APPROVED** [review]

File: boutdata/data.py (hunk 5 above)
Comment (johnomotani) [reply boutdata/data.py:1307]: Those probably would be good targets for refactoring, although I'm not sure how useful `BoutOutputs.redistribute` really is - redistributing `BOUT.dmp.*.nc` files is probably only useful if you want to append to them, but it's probably better to just squash the old dump files and new dump files separately now that `squashoutput` exists. It'd be more useful for `BoutOutputs` to be able to create redistributed restart files from the data in dump files, but `xBOUT` can already do that. Agree it's a thing for a separate PR though!

**Pull request merged**